sdk/servicebus/service-bus/src/core/batchingReceiver.ts

```diff
@@ -211,14 +213,19 @@ export class BatchingReceiver extends MessageReceiver {
 
       // Action to be performed on the "message" event.
       const onReceiveMessage: OnAmqpEventAsPromise = async (context: EventContext) => {
-        this.resetTimerOnNewMessageReceived();
+        if (brokeredMessages.length === 0) {
+          // We'll now remove the old timer (which was the overall `maxWaitTimeMs` timer)
+          // and replace it with another timer that is a (probably) much shorter interval.
+          //
+          // This allows the user to get access to received messages earlier and also gives us
+          // a chance to have fewer messages internally that could get lost if the user's
+          // app crashes in receiveAndDelete mode.
+          if (totalWaitTimer) clearTimeout(totalWaitTimer);
+          totalWaitTimer = setTimeout(actionAfterWaitTimeout, maxTimeAfterFirstMessageMs);
```

Comment:
This makes `receive` independent of the `maxMessageCount` and `maxWaitTime` for higher values.

Say, the first message is received in 3s, got 10 messages in the next one second. We'll be returning 11 messages right away.

For example, 
`(maxMessageCount = 1000, maxWaitTime = 100s)` becomes equivalent to 
`(maxMessageCount = 10000, maxWaitTime = 1000s)`

Should we do better at respecting the user-provided limits?

Reply:
It is a concern but this is the "core" of the current .NET algorithm. It favors the "fast" return of messages over waiting the assigned amount of time.

So the 'maxWaitTimeInMs' really represents "max time we'll wait for the first message".